packages/xo-server/src/xo-mixins/metadata-backups.js

```diff
@@ -0,0 +1,263 @@
+import defer from 'golike-defer'
+import { fromEvent, ignoreErrors } from 'promise-toolbox'
+
+import { type Xapi } from '../xapi'
+import {
+  safeDateFormat,
+  type SimpleIdPattern,
+  unboxIdsFromPattern,
+} from '../utils'
+
+import { type Executor, type Job } from './jobs'
+import { type Schedule } from './scheduling'
+
+const METADATA_BACKUP_JOB_TYPE = 'metadataBackup'
+
+type Settings = {|
+  retentionXoMetadata?: number,
+  retentionPoolMetadata?: number,
+|}
+
+type MetadataBackupJob = {
+  ...$Exact<Job>,
+  pools?: SimpleIdPattern,
+  remotes: SimpleIdPattern,
+  settings: $Dict<Settings>,
+  type: METADATA_BACKUP_JOB_TYPE,
+  xoMetadata?: boolean,
+}
+
+// File structure on remotes:
+//
+// <remote>
+// ├─ xo-config-backups
+// │  └─ <schedule ID>
+// │     └─ <YYYYMMDD>T<HHmmss>
+// │        ├─ metadata.json
+// │        └─ data.json
+// └─ xo-pool-metadata-backups
+//    └─ <schedule ID>
+//       └─ <pool UUID>
+//          └─ <YYYYMMDD>T<HHmmss>
+//             ├─ metadata.json
+//             └─ data
+
+export default class metadataBackup {
+  _app: {
+    createJob: (
+      $Diff<MetadataBackupJob, {| id: string |}>
+    ) => Promise<MetadataBackupJob>,
+    createSchedule: ($Diff<Schedule, {| id: string |}>) => Promise<Schedule>,
+    deleteSchedule: (id: string) => Promise<void>,
+    getXapi: (id: string) => Xapi,
+    getJob: (
+      id: string,
+      ?METADATA_BACKUP_JOB_TYPE
+    ) => Promise<MetadataBackupJob>,
+    updateJob: (
+      $Shape<MetadataBackupJob>,
+      ?boolean
+    ) => Promise<MetadataBackupJob>,
+    removeJob: (id: string) => Promise<void>,
+  }
+
+  constructor(app: any) {
+    this._app = app
+    app.on('start', () => {
+      app.registerJobExecutor(
+        METADATA_BACKUP_JOB_TYPE,
+        this._executor.bind(this)
+      )
+    })
+  }
+
+  async _executor({ cancelToken, job: job_, schedule }): Executor {
+    if (schedule === undefined) {
+      throw new Error('backup job cannot run without a schedule')
+    }
+
+    const job: MetadataBackupJob = (job_: any)
+    const remoteIds = unboxIdsFromPattern(job.remotes)
+    if (remoteIds.length === 0) {
+      throw new Error('metadata backup job cannot run without remotes')
+    }
+
+    const poolIds = unboxIdsFromPattern(job.pools)
+    const isEmptyPools = poolIds.length === 0
+    if (!job.xoMetadata && isEmptyPools) {
+      throw new Error('no metadata mode found')
+    }
+
+    const app = this._app
+    const { retentionXoMetadata, retentionPoolMetadata } =
+      job?.settings[schedule.id] || {}
+
+    const timestamp = safeDateFormat(Date.now())
+    const files = []
+    if (job.xoMetadata && retentionXoMetadata > 0) {
+      const xoMetadataDir = `xo-config-backups/${schedule.id}`
+      const dir = `${xoMetadataDir}/${timestamp}`
+
+      const data = JSON.stringify(await app.exportConfig(), null, 2)
+      const fileName = `${dir}/data.json`
+
+      const metadata = JSON.stringify(
+        {
+          jobId: job.id,
```

Comment:
Do you think this is enough?

How do you imagine the restore?

Reply:
I don't see other info to add, can you please enlighten me ?